### ISSUE TITLE
Environment integration, integration tests by nodejs and different browsers

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/o3#62]
https://github.com/inf3rno/dataflower/issues/26

**Comment by inf3rno** [comment]
Especially the different StackStringParsers should be tested in different environments. So I might need to move them to an adapter folder into separate files.

I'll end the lib with node.js after that I'll write dataflower and some server side applications, and only after that will I implement this, so it is not a priority.

**Comment by inf3rno** [comment]
https://github.com/inf3rno/dataflower/issues/59
Some info about stack string in different environments.

**Comment by inf3rno** [comment]
There should be separate env files for every supported environment. They should contain the StackStringParser implementation and they might contain polyfills like WeakMap, WeakSet, Function.prototype.bind, etc...

**Comment by inf3rno** [comment]
This should implement global error handling and error class injector as well. #57

**Comment by inf3rno** [comment]
- having custom Error classes, so I will be able to use instanceof in a catch block
 - v8 style stack frame in every supported environment (latest: node, chrome, firefox, msie, opera), the stack string should be generated by a custom method in the UserError class and in its sub-classes
 - handling uncaught exception and rejected future in v8 style in the browser as well, so instead of using window.blala, we use process.blabla everywhere
 - setting the stack length if possible (I think this is only possible by v8, but I am not sure)

sadly we cannot make a polyfill of Error.captureStackTrace, other than v8 browsers support this kind of features only partially, currently the adapter has a StackFrameExtractor.getFrameBeforeCall(fn) method, but in a non v8 environment I'll have to use a getFramesShifted(cnt) or something like that

nodejs has something like error containers. I forgot what's their name, but they can catch unhandled errors of different modules or objects. I am not sure about the logic, but it is a great feature, and I might be able to make a browser polyfill.

**Comment by inf3rno** [comment]
I found it, it is called domains: https://nodejs.org/api/domain.html . It is deprecated, maybe there is a better solution? I'll investigate this.

**Comment by inf3rno** [comment]
By classes there is nothing to tell about integration. Only ES5 environments will be supported, so I'll need an environment check to ensure this. By ES5- environments, enumeration and many other things work differently. The currently classes might be able to work in these environments, but if you start to build on this lib things like custom toString() methods, etc... then you will be in big trouble in ES5- environments.

**Comment by inf3rno** [comment]
Domain is not that great as I thought. It is better to use futures with catch to handle async exceptions than putting them under a domain. Afaik. domains are about running a function in an environment, which can catch oth the sync and async exceptions coming from that function. This is easy, you need to check whether the function is amongst the callers or not. Sadly this is not possible in most of the browsers, since we don't have v8 stack frames or arguments.callee.caller. So this won't be implemented in the adapter. I'll add something like this to dataflower, but it will be able to catch exceptions from a subgraph, or something like that, so it will have completely different logic.

**Comment by inf3rno** [comment]
I added chrome testing with karma. It only tests the class lib and the error.stack. The process.on() and the Error.stackTraceLimit is not tested. I am not sure how to write tests for those. As far as I understand mocha puts everything in try-catch blocks or uses process.on(...) to catch the exceptions. Both interferes with integration testing of these features.

**Comment by inf3rno** [comment]
Phantom, Chrome and Node are supported. The code is not perfect, but I want to add Firefox, Msie and Opera first. I can fix the bugs later.

**Comment by inf3rno** [comment]
Closing in favor of #83 #84
